tests/test_pandas.py

```diff
@@ -574,4 +573,493 @@ def test_cut() -> None:
+                ls,
+                rs,
+                how="inner",
+                left_index=True,
+                right_index=True,
+                suffixes=["_1", "_2"],
+            ),
+            pd.DataFrame,
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.merge(
+                ls,
+                rs,
+                how="inner",
+                left_index=True,
+                right_index=True,
+                suffixes=["_1", None],
+            ),
+            pd.DataFrame,
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.merge(
+                ls,
+                rs,
+                how="inner",
+                left_index=True,
+                right_index=True,
+                suffixes=("_1", None),
+            ),
+            pd.DataFrame,
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.merge(
+                ls,
+                rs,
+                how="inner",
+                left_index=True,
+                right_index=True,
+                suffixes=(None, "_2"),
+            ),
+            pd.DataFrame,
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.merge(
+                ls,
+                rs,
+                how="inner",
+                left_index=True,
+                right_index=True,
+                suffixes=("_1", "_2"),
+            ),
+            pd.DataFrame,
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.merge(
+                ls, rs, how="inner", left_index=True, right_index=True, indicator=True
+            ),
+            pd.DataFrame,
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.merge(lf, rs, left_on="left", right_on="right"), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.merge(ls, rf, left_on="left", right_on="right"), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.merge(lf, rf, left_on="left", right_on="right"), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.merge(lf, rf, left_on=["left"], right_on=["right"]), pd.DataFrame
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.merge(lf, rf, left_index=True, right_index=True), pd.DataFrame),
+        pd.DataFrame,
+    )
+
+
+def test_merge_ordered() -> None:
+    ls = pd.Series([1, 2, 3, 4], index=[1, 2, 3, 4], name="left")
+    rs = pd.Series([3, 4, 5, 6], index=[3, 4, 5, 6], name="right")
+    lf = pd.DataFrame(
+        [[1, 2, 3], [3, 4, 5], [5, 6, 7], [7, 8, 9]],
+        index=[1, 2, 3, 4],
+        columns=["a", "b", "c"],
+    )
+    rf = pd.DataFrame(pd.Series([3, 4, 5, 6], index=[3, 4, 5, 6], name="b"))
+
+    check(
+        assert_type(
+            pd.merge_ordered(ls, rs, left_on="left", right_on="right"), pd.DataFrame
```

Comment:
you have overloads for merging series with dataframe, dataframe with series, so maybe test those as well.

And you could add a test with `if TYPE_CHECKING_INVALID_USAGE` and a `#type: ignore` to see that we raise a typing error if you try to do `pd.merge_ordered(pd.Series(), pd.Series())`

Reply:
`Series, Series` is OK.  The overload are to force the values of `left_by` and `right_by`  ro be `None` when at least one of the inputs is a Series.

Reply:
I added something along these lines.